libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/offroute/OffRouteDetector.java

```diff
@@ -226,24 +224,28 @@ private static boolean movingAwayFromManeuver(RouteProgress routeProgress,
     LineString remainingStepLineString = TurfMisc.lineSlice(userPointOnStep, maneuverPoint, stepLineString);
     double userDistanceToManeuver = TurfMeasurement.length(remainingStepLineString, TurfConstants.UNIT_METERS);
 
-    boolean hasDistances = !distancesAwayFromManeuver.isEmpty();
-    boolean validOffRouteDistanceTraveled = hasDistances && distancesAwayFromManeuver.peekLast()
-      - distancesAwayFromManeuver.peekFirst() < MINIMUM_BACKUP_DISTANCE_FOR_OFF_ROUTE;
-    boolean exceedsManeuverDistancesThreshold = validOffRouteDistanceTraveled
-      && distancesAwayFromManeuver.size() >= 3;
-
-    if (exceedsManeuverDistancesThreshold) {
-      // User's moving away from maneuver position, thus offRoute.
-      return true;
-    }
     if (distancesAwayFromManeuver.isEmpty()) {
-      distancesAwayFromManeuver.push((int) userDistanceToManeuver);
-    } else if (userDistanceToManeuver > distancesAwayFromManeuver.peek()) {
-      distancesAwayFromManeuver.push((int) userDistanceToManeuver);
-    } else {
-      // If we get a descending distance, reset the counter
+      // No move-away positions before, add the current one to history stack
+      distancesAwayFromManeuver.add((int) userDistanceToManeuver);
+    } else if ((int) userDistanceToManeuver > distancesAwayFromManeuver.getLast()) {
+      // If distance to maneuver increased (wrong way), add new position to history stack
+
+      if (distancesAwayFromManeuver.size() >= 3) {
+        // We replace the first one, in order to keep the history with the last one
+        distancesAwayFromManeuver.removeLast();
+      }
+      distancesAwayFromManeuver.add((int) userDistanceToManeuver);
+    } else if ((int) userDistanceToManeuver < distancesAwayFromManeuver.getLast()) {
+      // If distance to maneuver decreased (right way) clean history
       distancesAwayFromManeuver.clear();
     }
+
+    // Minimum 3 position updates in the wrong way are required before an off-route can occur
+    if (distancesAwayFromManeuver.size() >= 3) {
+      // Check for minimum distance traveled
+      return (distancesAwayFromManeuver.getLast() - distancesAwayFromManeuver.getFirst()) > MINIMUM_BACKUP_DISTANCE_FOR_OFF_ROUTE;
```

Comment:
We could think about using a value from MapboxNavigationOptions instead of a constant, so that implementers can change it? But since it was like that before, it would be fine for a follow up as well.